odbc/src/api/data.rs

```diff
@@ -0,0 +1,647 @@
+use crate::{errors::ODBCError, handles::definitions::MongoHandle};
+use bson::Bson;
+use chrono::{
+    offset::{TimeZone, Utc},
+    DateTime, Datelike, Timelike,
+};
+use mongo_odbc_core::util::Decimal128Plus;
+use odbc_sys::{CDataType, Date, Len, Pointer, Time, Timestamp};
+use odbc_sys::{Char, Integer, SmallInt, SqlReturn, WChar};
+use std::{cmp::min, mem::size_of, ptr::copy_nonoverlapping, str::FromStr};
+
+/// IntoCData is just used for adding methods to bson::Bson.
+trait IntoCData {
+    fn to_string(self) -> String;
+    fn to_f64(self) -> f64;
+    fn to_f32(self) -> f32;
+    fn to_i64(self) -> i64;
+    fn to_i32(self) -> i32;
+    fn to_bool(self) -> bool;
+    fn to_date(self) -> DateTime<Utc>;
+}
+
+impl IntoCData for Bson {
+    fn to_string(self) -> String {
+        match self {
+            Bson::String(s) => s,
+            // TODO SQL-1068 :we will have to test this manually because there is no way to create a Decimal128
+            // in a unit test at this time. We could load a bson file, but since Decimal128 support
+            // will be added soon, it's fine to wait on full support.
+            Bson::Decimal128(d) => {
+                format!("{{$numberDecimal: \"{}\"}}", d.to_formatted_string())
+            }
+            Bson::Array(_)
+            | Bson::Binary(_)
+            | Bson::Boolean(_)
+            | Bson::Double(_)
+            | Bson::DateTime(_)
+            | Bson::DbPointer(_)
+            | Bson::Document(_)
+            | Bson::JavaScriptCode(_)
+            | Bson::JavaScriptCodeWithScope(_)
+            | Bson::Int32(_)
+            | Bson::Int64(_)
+            | Bson::MaxKey
+            | Bson::MinKey
+            | Bson::ObjectId(_)
+            | Bson::RegularExpression(_)
+            | Bson::Symbol(_)
+            | Bson::Timestamp(_) => self.into_canonical_extjson().to_string(),
+            Bson::Null | Bson::Undefined => unreachable!(),
```

Comment:
This will be handled in the next ticket because I am not sure yet what is required by the ODBC standard. fwiw, currently we return null when we ask for a string value from null because of the NULL_DATA case at the beginning of the function.